src/Mvc/Mvc.Core/src/ModelBinding/Binders/KeyedServicesModelBinder.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using Microsoft.AspNetCore.Mvc.ModelBinding.Validation;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Binders;
+
+internal class KeyedServicesModelBinder : IModelBinder
+{
+    private readonly object _key;
+    private readonly bool _isOptional;
+
+    public KeyedServicesModelBinder(object key, bool isOptional)
+    {
+        _key = key ?? throw new ArgumentNullException(nameof(key));
+        _isOptional = isOptional;
+    }
+
+    public Task BindModelAsync(ModelBindingContext bindingContext)
+    {
+        var keyedServices = bindingContext.HttpContext.RequestServices as IKeyedServiceProvider;
```

Comment:
Is there any chance that this cast would fail?

Reply:
Yes, if the provider doesn't support keyed services.